src/main/java/soconnect/logic/autocomplete/AutocompleteManager.java

```diff
@@ -1,47 +1,202 @@
+        case SearchCommand.EMPTY_CONDITION:
+            filteredPersons.setPredicate(new ContactContainsAllKeywordsPredicate(argMultimap));
+            break;
+        case SearchCommand.OR_CONDITION:
+            filteredPersons.setPredicate(PREDICATE_SHOW_ALL_PERSONS);
+            break;
+        default:
+            filteredPersons.setPredicate(PREDICATE_SHOW_NO_PERSON);
+        }
+
+    }
+
+    @Override
+    public List<String> generateAutocompleteEntries(String argsWithoutLastPrefixArgument,
+                                                    String lastPrefixArgument) {
+        String autocompleteString = SearchCommand.COMMAND_WORD + argsWithoutLastPrefixArgument;
+
+        String[] prefixArgument = lastPrefixArgument.split("/", 2);
+        String prefix = prefixArgument[0] + "/";
+        String argument = prefixArgument[1];
+
+        switch (prefix) {
+        case INDICATOR_NAME:
+            return filteredPersons.stream()
+                    .filter(person -> startWithIgnoreCase(person.getName().fullName, argument))
+                    .limit(AUTOCOMPLETE_ENTRIES_LIMIT)
+                    .map(person -> autocompleteString + prefix + person.getName().fullName)
+                    .collect(Collectors.toList());
+        case INDICATOR_ADDRESS:
+            return filteredPersons.stream()
+                    .filter(person -> startWithIgnoreCase(person.getAddress().value, argument))
+                    .limit(AUTOCOMPLETE_ENTRIES_LIMIT)
+                    .map(person -> autocompleteString + prefix + person.getAddress().value)
+                    .collect(Collectors.toList());
+        case INDICATOR_EMAIL:
+            return filteredPersons.stream()
+                    .filter(person -> startWithIgnoreCase(person.getEmail().value, argument))
+                    .limit(AUTOCOMPLETE_ENTRIES_LIMIT)
+                    .map(person -> autocompleteString + prefix + person.getEmail().value)
+                    .collect(Collectors.toList());
+        case INDICATOR_PHONE:
+            return filteredPersons.stream()
+                    .filter(person -> startWithIgnoreCase(person.getPhone().value, argument))
+                    .limit(AUTOCOMPLETE_ENTRIES_LIMIT)
+                    .map(person -> autocompleteString + prefix + person.getPhone().value)
+                    .collect(Collectors.toList());
+        case INDICATOR_TAG:
+            return filteredPersons.stream()
+                    .map(person -> person.getTags().stream()
+                            .filter(tag -> startWithIgnoreCase(tag.tagName, argument))
+                            .map(tag -> autocompleteString + prefix + tag.tagName)
+                            .collect(Collectors.toList()))
+                    .flatMap(Collection::stream)
+                    .distinct()
+                    .limit(AUTOCOMPLETE_ENTRIES_LIMIT)
+                    .collect(Collectors.toList());
+        default:
+            return new ArrayList<>();
```

Comment:
The length is a bit concerning haha, but as long as you can defend yourself should be okay ;D

Reply:
Yup, I think it is too long, I will extract each of them out and make a private method for each indicator.

Reply:
The newest addition looks good!